In [69]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [70]:
# 2

file_path = "gmo_analysis_data.xlsx"
tr = pd.read_excel(file_path, sheet_name="total returns")
rf = pd.read_excel(file_path, sheet_name="risk-free rate")
signal = pd.read_excel(file_path, sheet_name="signals")
tr["date"] = pd.to_datetime(tr["date"])
rf["date"] = pd.to_datetime(rf["date"])
signal["date"] = pd.to_datetime(signal["date"])
tr.set_index("date", inplace=True)
rf.set_index("date", inplace=True)
signal.set_index("date", inplace=True) 

er = tr - rf.values
er.head()

SPY     GMWAX     GMGEX
date                                    
1996-12-31 -0.075002 -0.073804 -0.064710
1997-01-31  0.010316 -0.036735 -0.017022
1997-02-28 -0.042635 -0.029935 -0.039467
1997-03-31 -0.098941 -0.068372 -0.069661
1997-04-30  0.012038 -0.059061 -0.052330

In [71]:
FREQ = 12

def performance_stats(ret):
    """Compute mean, vol, Sharpe (annualized)."""
    mean_ann = ret.mean() * FREQ
    vol_ann = ret.std() * np.sqrt(FREQ)
    sharpe = mean_ann / vol_ann if vol_ann != 0 else np.nan
    return mean_ann, vol_ann, sharpe

def max_drawdown(total_ret):
    wealth = (1 + total_ret).cumprod()
    peak = wealth.cummax()
    dd = (wealth - peak) / peak
    return dd.min()

def tail_risk(ret, total_ret):
    min_ret = ret.min()
    var_5 = ret.quantile(0.05)
    mdd = max_drawdown(total_ret)
    return min_ret, var_5, mdd

def regression_stats(y, x):
    X = sm.add_constant(x)
    model = sm.OLS(y, X).fit()
    alpha_daily = model.params["const"]
    beta = model.params[x.name]
    r2 = model.rsquared
    alpha_ann = alpha_daily * FREQ
    return alpha_ann, beta, r2

In [72]:
# 2.1

samples = {
    "inception_to_2011": er.loc[: "2011-12-31"],
    "2012_to_present":   er.loc["2012-01-01": ],
    "inception_to_present": er
}

print("===== 2.1 PERFORMANCE (GMWAX) =====")
for name, sub in samples.items():
    m, v, s = performance_stats(sub["GMWAX"])
    print(f"\n--- {name} ---")
    print(f"Mean (ann):   {m:.6f}")
    print(f"Vol (ann):    {v:.6f}")
    print(f"Sharpe:       {s:.6f}")

===== 2.1 PERFORMANCE (GMWAX) =====

--- inception_to_2011 ---
Mean (ann):   -0.265291
Vol (ann):    0.131867
Sharpe:       -2.011809

--- 2012_to_present ---
Mean (ann):   -0.123303
Vol (ann):    0.109588
Sharpe:       -1.125154

--- inception_to_present ---
Mean (ann):   -0.197366
Vol (ann):    0.123263
Sharpe:       -1.601179


<span style="color: blue;">

ANS:

The mean became less negative in sub_period 2, while volatility also become less, the Sharpe ratio improve in the sub_period 2, from -2.0 to -1.1.

In [73]:
#2.2

print("\n===== 2.2 TAIL RISK (GMWAX) =====")
for name, sub in samples.items():
    min_ret, var5, mdd = tail_risk(
        sub["GMWAX"],            # excess return
        tr.loc[sub.index]["GMWAX"]    # total return
    )
    print(f"\n--- {name} ---")
    print(f"Min return:   {min_ret:.6f}")
    print(f"VaR-5%:       {var5:.6f}")
    print(f"Max drawdown: {mdd:.6f}")


===== 2.2 TAIL RISK (GMWAX) =====

--- inception_to_2011 ---
Min return:   -0.193379
VaR-5%:       -0.083767
Max drawdown: -0.293614

--- 2012_to_present ---
Min return:   -0.115577
VaR-5%:       -0.071114
Max drawdown: -0.216795

--- inception_to_present ---
Min return:   -0.193379
VaR-5%:       -0.077006
Max drawdown: -0.293614


<span style="color: blue;">

ANS:

(a) The tail-risk seems high, as Var-5% of monthly data is 7-8% through the sample, and the minimum return is 19% in all the sample, and maximum drawdown is 29.4% in all the sample.

(b) Yes, the second sub-sample seems less ricky in term of tail risk. But it is still quite risky as the Var-5% is 7.1%, maximum drawdown is 22%, and minimum return per month is 11.6%.

In [74]:
# 2.3

print("\n===== 2.3 MARKET EXPOSURE (GMWAX) =====")
for name, sub in samples.items():
    alpha, beta, r2 = regression_stats(
        sub["GMWAX"],     # y
        sub["SPY"]  # x = SPY excess
    )
    print(f"\n--- {name} ---")
    print(f"Alpha (ann):  {alpha:.6f}")
    print(f"Beta:         {beta:.6f}")
    print(f"R²:           {r2:.6f}")



===== 2.3 MARKET EXPOSURE (GMWAX) =====

--- inception_to_2011 ---
Alpha (ann):  -0.094369
Beta:         0.619535
R²:           0.687802

--- 2012_to_present ---
Alpha (ann):  -0.099708
Beta:         0.629431
R²:           0.769300

--- inception_to_present ---
Alpha (ann):  -0.096637
Beta:         0.622398
R²:           0.727059


<span style="color: blue;">

ANS:

GMWAX has around 0.6 Beta, I would not consider it low-beta strategy, and this didn't change across subsamples.

GMWAX does not provide alpha, it has negative alpha across both subsamples.

In [75]:
# 2.4

def analyze_one_fund(fund_name):
    print(f"\n========== RESULTS FOR {fund_name} ==========")

    # 2.1 Performance
    print("\n--- 2.1 PERFORMANCE ---")
    for name, sub in samples.items():
        m, v, s = performance_stats(sub[fund_name])
        print(f"\n{name}")
        print(f"Mean (ann):   {m:.6f}")
        print(f"Vol (ann):    {v:.6f}")
        print(f"Sharpe:       {s:.6f}")

    # 2.2 Tail Risk
    print("\n--- 2.2 TAIL RISK ---")
    for name, sub in samples.items():
        min_ret, var5, mdd = tail_risk(
            sub[fund_name],
            tr.loc[sub.index][fund_name]
        )
        print(f"\n{name}")
        print(f"Min return:   {min_ret:.6f}")
        print(f"VaR-5%:       {var5:.6f}")
        print(f"Max drawdown: {mdd:.6f}")

    # 2.3 Market Exposure Regression
    print("\n--- 2.3 REGRESSION ---")
    for name, sub in samples.items():
        alpha, beta, r2 = regression_stats(
            sub[fund_name],
            er.loc[sub.index]["SPY"]
        )
        print(f"\n{name}")
        print(f"Alpha (ann):  {alpha:.6f}")
        print(f"Beta:         {beta:.6f}")
        print(f"R²:           {r2:.6f}")

# Run comparison
analyze_one_fund("GMWAX")
analyze_one_fund("GMGEX")


========== RESULTS FOR GMWAX ==========

--- 2.1 PERFORMANCE ---

inception_to_2011
Mean (ann):   -0.265291
Vol (ann):    0.131867
Sharpe:       -2.011809

2012_to_present
Mean (ann):   -0.123303
Vol (ann):    0.109588
Sharpe:       -1.125154

inception_to_present
Mean (ann):   -0.197366
Vol (ann):    0.123263
Sharpe:       -1.601179

--- 2.2 TAIL RISK ---

inception_to_2011
Min return:   -0.193379
VaR-5%:       -0.083767
Max drawdown: -0.293614

2012_to_present
Min return:   -0.115577
VaR-5%:       -0.071114
Max drawdown: -0.216795

inception_to_present
Min return:   -0.193379
VaR-5%:       -0.077006
Max drawdown: -0.293614

--- 2.3 REGRESSION ---

inception_to_2011
Alpha (ann):  -0.094369
Beta:         0.619535
R²:           0.687802

2012_to_present
Alpha (ann):  -0.099708
Beta:         0.629431
R²:           0.769300

inception_to_present
Alpha (ann):  -0.096637
Beta:         0.622398
R²:           0.727059

========== RESULTS FOR GMGEX ==========

--- 2.1 PERFORMANCE ---

incepti

<span style="color: blue;">

ANS:

Performance Comparison

GMGEX underperforms GMWAX consistently, showing lower annualized returns and significantly higher volatility.

Both funds have negative Sharpe ratios, but GMGEX’s Sharpe is worse, especially in the post-2012 period.

Tail-Risk Comparison

GMGEX exhibits substantially higher tail risk, including a maximum single-day loss of –66% and maximum drawdowns deeper than –70%.

GMWAX’s worst outcomes are far milder, with drawdowns around –30%.

Market Exposure

GMWAX has a stable low beta (~0.62), consistent with a lower-risk profile.

GMGEX has a higher beta (~0.80) and less stable market exposure (R² drops to ~0.28 after 2012).

Both funds deliver negative alpha, but GMGEX’s alpha deteriorates more severely.

Overall Conclusion

GMWAX is a lower-risk, lower-beta, but low-return strategy.
GMGEX carries much higher risk, larger drawdowns, unstable behavior, and even worse performance.


In [76]:
# 3

df = tr.join(rf, how="inner", rsuffix="_rf").join(signal, how="inner")
df["er"] = df["SPY"] - df["TBill 3M"]

df["dp_lag"] = df["SPX D/P"].shift(1)
df["ep_lag"] = df["SPX E/P"].shift(1)
df["y10_lag"] = df["T-Note 10YR"].shift(1)

df = df.dropna()

def regress(y, X):
    X = sm.add_constant(X)
    return sm.OLS(y, X).fit()

def compute_max_dd(series):
    cum = (1 + series).cumprod()
    peak = cum.cummax()
    dd = (cum - peak) / peak
    return dd.min()

def information_ratio(alpha, residuals):
    return alpha / residuals.std()

In [77]:
# 3.1

y = df["SPY"]

model_dp = regress(y, df[["dp_lag"]])
model_ep = regress(y, df[["ep_lag"]])
model_three = regress(y, df[["dp_lag", "ep_lag", "y10_lag"]])

print("===== REGRESSIONS =====")
print("(1) dp-lag")
print("R²:", round(model_dp.rsquared, 6))
print(model_dp.params.round(6))

print("\n(2) ep-lag")
print("R²:", round(model_ep.rsquared, 6))
print(model_ep.params.round(6))

print("\n(3) dp + ep + y10")
print("R²:", round(model_three.rsquared, 6))
print(model_three.params.round(6))

===== REGRESSIONS =====
(1) dp-lag
R²: 0.007263
const    -0.007793
dp_lag    0.928617
dtype: float64

(2) ep-lag
R²: 0.004827
const    -0.004074
ep_lag    0.240449
dtype: float64

(3) dp + ep + y10
R²: 0.008638
const     -0.002655
dp_lag     0.445540
ep_lag     0.142754
y10_lag   -0.117296
dtype: float64


In [78]:
# 3.2

def build_strategy(model, Xlag):
    Xlag = sm.add_constant(Xlag)
    forecast = model.predict(Xlag)      # E_t[er_{t+1}]
    weight = forecast * 100                  # w_t = forecast
    strategy_ret = weight * df["SPY"]    # realized return
    return strategy_ret, forecast

ret_dp, fc_dp = build_strategy(model_dp, df[["dp_lag"]])
ret_ep, fc_ep = build_strategy(model_ep, df[["ep_lag"]])
ret_three, fc_three = build_strategy(model_three, df[["dp_lag", "ep_lag", "y10_lag"]])

def perf_stats(strategy_ret):
    mean = strategy_ret.mean()
    vol = strategy_ret.std()
    sharpe = mean / vol

    maxdd = compute_max_dd(strategy_ret)

    X = sm.add_constant(df.loc[strategy_ret.index, "SPY"])
    mkt_model = sm.OLS(strategy_ret, X).fit()
    alpha = mkt_model.params["const"]
    beta = mkt_model.params["SPY"]
    ir = information_ratio(alpha, mkt_model.resid)

    return {
        "mean": round(mean,6),
        "vol": round(vol,6),
        "sharpe": round(sharpe,6),
        "maxdd": round(maxdd,6),
        "alpha": round(alpha,6),
        "beta": round(beta,6),
        "IR": round(ir,6),
    }

stats_dp = perf_stats(ret_dp)
stats_ep = perf_stats(ret_ep)
stats_three = perf_stats(ret_three)

print("\n===== STRATEGY PERFORMANCE =====")
print("DP:", stats_dp)
print("EP:", stats_ep)
print("Three:", stats_three)


===== STRATEGY PERFORMANCE =====
DP: {'mean': np.float64(0.00932), 'vol': np.float64(0.048565), 'sharpe': np.float64(0.191919), 'maxdd': np.float64(-0.698615), 'alpha': np.float64(0.000711), 'beta': np.float64(0.96872), 'IR': np.float64(0.031287)}
EP: {'mean': np.float64(0.008844), 'vol': np.float64(0.044907), 'sharpe': np.float64(0.196935), 'maxdd': np.float64(-0.616841), 'alpha': np.float64(0.000416), 'beta': np.float64(0.948248), 'IR': np.float64(0.026243)}
Three: {'mean': np.float64(0.00959), 'vol': np.float64(0.048226), 'sharpe': np.float64(0.198846), 'maxdd': np.float64(-0.666069), 'alpha': np.float64(0.001045), 'beta': np.float64(0.96144), 'IR': np.float64(0.046211)}


In [79]:
# 3.3

VaR = {
    "dp": round(ret_dp.quantile(0.05),6),
    "ep": round(ret_ep.quantile(0.05),6),
    "three": round(ret_three.quantile(0.05),6),
    "market": round(df["SPY"].quantile(0.05),6),
}

print("\n===== VAR(5%) =====")
print(VaR)


df_sub = df.loc["2000-01-01":"2011-12-31"]

def underperf(ret):
    return round((ret.loc[df_sub.index] - df_sub["TBill 3M"]).mean(),6)

print("\n===== UNDERPERFORMANCE VS RF 2000–2011 =====")
print("dp:", underperf(ret_dp))
print("ep:", underperf(ret_ep))
print("three:", underperf(ret_three))

print("\n===== NEGATIVE RISK PREMIUM COUNTS =====")
print("dp:", (fc_dp < 0).sum())
print("ep:", (fc_ep < 0).sum())
print("three:", (fc_three < 0).sum())


===== VAR(5%) =====
{'dp': np.float64(-0.061545), 'ep': np.float64(-0.064144), 'three': np.float64(-0.065216), 'market': np.float64(-0.074428)}

===== UNDERPERFORMANCE VS RF 2000–2011 =====
dp: -0.018652
ep: -0.019573
three: -0.018332

===== NEGATIVE RISK PREMIUM COUNTS =====
dp: 0
ep: 0
three: 1


<span style="color: blue;">

ANS:

Do you believe the dynamic strategy takes on extra risk?
I don't believe so, and if we look at the var-5%, the dynamic strategy perform better than SPY. 
However, the three factor dynamic strategy shows a 0.1% alpha (for monthly data). This may mean it take extra risk base on markey efficiency theory.

In [80]:
# 4.1

y = df["SPY"]
X = df[["dp_lag", "ep_lag"]]
X = sm.add_constant(X)


start_window = 61                     # 10 years of monthly data
T = len(df)

oos_forecast = []
oos_actual = []
oos_null = []

for t in range(start_window, T - 1):

    # Rolling window: up to t
    y_train = y.iloc[:t]
    X_train = X.iloc[:t]

    # Estimate model
    model = sm.OLS(y_train, X_train).fit()

    # Forecast for t+1
    Xt1 = X.iloc[t:t+1]
    f = float(model.predict(Xt1))

    # Actual t+1
    r_actual = float(y.iloc[t+1])

    # Null forecast: mean of y_train
    f_null = float(y_train.mean())

    oos_forecast.append(f)
    oos_actual.append(r_actual)
    oos_null.append(f_null)

# Convert to Series
oos_forecast = pd.Series(oos_forecast, index=y.index[start_window+1:])
oos_actual = pd.Series(oos_actual, index=y.index[start_window+1:])
oos_null = pd.Series(oos_null, index=y.index[start_window+1:])

errors_model = oos_actual - oos_forecast
errors_null = oos_actual - oos_null

RSS_model = np.sum(errors_model**2)
RSS_null = np.sum(errors_null**2)

R2_oos = 1 - RSS_model / RSS_null

print("===== OOS R^2 =====")
print("R2_out_of_sample =", round(R2_oos, 6))
print("Positive R2?", R2_oos > 0)


===== OOS R^2 =====
R2_out_of_sample = -0.065356
Positive R2? False


C:\Users\User\AppData\Local\Temp\ipykernel_13264\631002566.py:26: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  f = float(model.predict(Xt1))


<span style="color: blue;">

ANS:

No, it actually has a negative number. It shows that the out-od -sample prediction is weak, and there might be overfitting problem.

In [81]:
# 4.2

y = df["SPY"]      # market return, consistent with your 3.2
X_dp = df[["dp_lag"]]
X_ep = df[["ep_lag"]]
X_three = df[["dp_lag", "ep_lag", "y10_lag"]]
X_dp = sm.add_constant(X_dp)
X_ep = sm.add_constant(X_ep)
X_three = sm.add_constant(X_three)

start_window = 60
T = len(df)

# Containers for OOS forecasts
oos_dp = []
oos_ep = []
oos_three = []

# Actual SPY returns at t+1
oos_actual = []

# ==========================================================
# Rolling OOS Loop
# ==========================================================

for t in range(start_window, T - 1):

    # OOS actual return (t+1)
    r_tp1 = float(y.iloc[t+1])
    oos_actual.append(r_tp1)

    # --------------------------
    # Model 1: dp only
    # --------------------------
    y_train = y.iloc[:t]
    X_train_dp = X_dp.iloc[:t]

    model_dp = sm.OLS(y_train, X_train_dp).fit()

    Xt1_dp = X_dp.iloc[t:t+1]
    f_dp = float(model_dp.predict(Xt1_dp))
    oos_dp.append(f_dp)

    # --------------------------
    # Model 2: ep only
    # --------------------------
    X_train_ep = X_ep.iloc[:t]

    model_ep = sm.OLS(y_train, X_train_ep).fit()

    Xt1_ep = X_ep.iloc[t:t+1]
    f_ep = float(model_ep.predict(Xt1_ep))
    oos_ep.append(f_ep)

    # --------------------------
    # Model 3: three regressors
    # --------------------------
    X_train_three = X_three.iloc[:t]

    model_three = sm.OLS(y_train, X_train_three).fit()

    Xt1_three = X_three.iloc[t:t+1]
    f_three = float(model_three.predict(Xt1_three))
    oos_three.append(f_three)


# Convert to Series with correct dates
index_oos = y.index[start_window+1:]

oos_dp = pd.Series(oos_dp, index=index_oos)
oos_ep = pd.Series(oos_ep, index=index_oos)
oos_three = pd.Series(oos_three, index=index_oos)
oos_actual = pd.Series(oos_actual, index=index_oos)

def build_oos_strategy(oos_forecast, oos_actual):
    weight = oos_forecast * 100
    strat_ret = weight * oos_actual
    return strat_ret

oos_ret_dp = build_oos_strategy(oos_dp, oos_actual)
oos_ret_ep = build_oos_strategy(oos_ep, oos_actual)
oos_ret_three = build_oos_strategy(oos_three, oos_actual)

stats_oos_dp = perf_stats(oos_ret_dp)
stats_oos_ep = perf_stats(oos_ret_ep)
stats_oos_three = perf_stats(oos_ret_three)

print("\n===== OOS Strategy Performance =====")
print("OOS dp:", stats_oos_dp)
print("OOS ep:", stats_oos_ep)
print("OOS three:", stats_oos_three)

C:\Users\User\AppData\Local\Temp\ipykernel_13264\3416657851.py:41: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  f_dp = float(model_dp.predict(Xt1_dp))
C:\Users\User\AppData\Local\Temp\ipykernel_13264\3416657851.py:52: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  f_ep = float(model_ep.predict(Xt1_ep))
C:\Users\User\AppData\Local\Temp\ipykernel_13264\3416657851.py:63: FutureWarning: Calling float on a single element Series is deprecated and will raise a TypeError in the future. Use float(ser.iloc[0]) instead
  f_three = float(model_three.predict(Xt1_three))



===== OOS Strategy Performance =====
OOS dp: {'mean': np.float64(0.005505), 'vol': np.float64(0.055733), 'sharpe': np.float64(0.098773), 'maxdd': np.float64(-0.774654), 'alpha': np.float64(-0.002336), 'beta': np.float64(0.891687), 'IR': np.float64(-0.057666)}
OOS ep: {'mean': np.float64(0.005425), 'vol': np.float64(0.032577), 'sharpe': np.float64(0.166541), 'maxdd': np.float64(-0.563231), 'alpha': np.float64(-0.000587), 'beta': np.float64(0.683786), 'IR': np.float64(-0.041573)}
OOS three: {'mean': np.float64(0.002572), 'vol': np.float64(0.048844), 'sharpe': np.float64(0.052666), 'maxdd': np.float64(-0.751863), 'alpha': np.float64(-0.003906), 'beta': np.float64(0.736756), 'IR': np.float64(-0.104943)}


In [82]:
comparison = pd.DataFrame({
    "InSample_dp": stats_dp,
    "OOS_dp": stats_oos_dp,
    "InSample_ep": stats_ep,
    "OOS_ep": stats_oos_ep,
    "InSample_three": stats_three,
    "OOS_three": stats_oos_three
})

# Make table cleaner: transpose
comparison = comparison.T

print("\n===== In-Sample vs OOS Strategy Comparison =====")
display(comparison)


===== In-Sample vs OOS Strategy Comparison =====


mean       vol    sharpe     maxdd     alpha      beta  \
InSample_dp     0.009320  0.048565  0.191919 -0.698615  0.000711  0.968720   
OOS_dp          0.005505  0.055733  0.098773 -0.774654 -0.002336  0.891687   
InSample_ep     0.008844  0.044907  0.196935 -0.616841  0.000416  0.948248   
OOS_ep          0.005425  0.032577  0.166541 -0.563231 -0.000587  0.683786   
InSample_three  0.009590  0.048226  0.198846 -0.666069  0.001045  0.961440   
OOS_three       0.002572  0.048844  0.052666 -0.751863 -0.003906  0.736756   

                      IR  
InSample_dp     0.031287  
OOS_dp         -0.057666  
InSample_ep     0.026243  
OOS_ep         -0.041573  
InSample_three  0.046211  
OOS_three      -0.104943

<span style="color: blue;">

ANS:

Across all three predictive models (dp, ep, and the three-variable model), performance deteriorates substantially when moving from in-sample (IS) to out-of-sample (OOS) evaluation. Mean returns and Sharpe ratios drop sharply in the OOS period, while maximum drawdowns become deeper and more volatile. Although the three-variable model appears strongest in-sample, it suffers the largest OOS performance decay, indicating clear overfitting.

All models exhibit a decline in alpha—turning negative OOS—suggesting the absence of persistent excess returns. Beta values also fall meaningfully, showing that OOS exposures become unstable and less aligned with the market. Finally, Information Ratios turn negative for every model, confirming that the predictive signals lack OOS informational content.

Overall, the results show that these valuation-based predictors (dp, ep, and the combined model) possess weak and unstable forecasting power in real-time applications, despite looking favorable in-sample.


In [83]:
# 4.3

VaR_insample = {
    "dp": round(ret_dp.quantile(0.05),6),
    "ep": round(ret_ep.quantile(0.05),6),
    "three": round(ret_three.quantile(0.05),6),
}

VaR_oos = {
    "dp": round(oos_ret_dp.quantile(0.05),6),
    "ep": round(oos_ret_ep.quantile(0.05),6),
    "three": round(oos_ret_three.quantile(0.05),6),
}

print("\n===== VaR(5%) In-Sample =====")
print(VaR_insample)

print("\n===== VaR(5%) OOS =====")
print(VaR_oos)


===== VaR(5%) In-Sample =====
{'dp': np.float64(-0.061545), 'ep': np.float64(-0.064144), 'three': np.float64(-0.065216)}

===== VaR(5%) OOS =====
{'dp': np.float64(-0.06332), 'ep': np.float64(-0.047868), 'three': np.float64(-0.074456)}


In [88]:
df_sub = df.loc["2000-01-01":"2011-12-31"]

def underperf(ret):
    idx = ret.index.intersection(df_sub.index)
    return round((ret.loc[idx] - df_sub["TBill 3M"].loc[idx]).mean(), 6)

under_insample = {
    "dp": underperf(ret_dp),
    "ep": underperf(ret_ep),
    "three": underperf(ret_three)
}

under_oos = {
    "dp": underperf(oos_ret_dp),
    "ep": underperf(oos_ret_ep),
    "three": underperf(oos_ret_three)
}

print("\n===== UNDERPERFORMANCE VS RF 2000–2011 (In-Sample) =====")
print(under_insample)

print("\n===== UNDERPERFORMANCE VS RF 2000–2011 (OOS) =====")
print(under_oos)


===== UNDERPERFORMANCE VS RF 2000–2011 (In-Sample) =====
{'dp': np.float64(-0.018652), 'ep': np.float64(-0.019573), 'three': np.float64(-0.018332)}

===== UNDERPERFORMANCE VS RF 2000–2011 (OOS) =====
{'dp': np.float64(-0.018093), 'ep': np.float64(-0.015519), 'three': np.float64(-0.022612)}


In [89]:
neg_insample = {
    "dp": (fc_dp < 0).sum(),
    "ep": (fc_ep < 0).sum(),
    "three": (fc_three < 0).sum()
}

neg_oos = {
    "dp": (oos_dp < 0).sum(),
    "ep": (oos_ep < 0).sum(),
    "three": (oos_three < 0).sum()
}

print("\n===== NEGATIVE RISK PREMIUM COUNTS (In-Sample) =====")
print(neg_insample)

print("\n===== NEGATIVE RISK PREMIUM COUNTS (OOS) =====")
print(neg_oos)


===== NEGATIVE RISK PREMIUM COUNTS (In-Sample) =====
{'dp': np.int64(0), 'ep': np.int64(0), 'three': np.int64(1)}

===== NEGATIVE RISK PREMIUM COUNTS (OOS) =====
{'dp': np.int64(5), 'ep': np.int64(0), 'three': np.int64(8)}


In [90]:
table_33 = pd.DataFrame({
    "InSample_VaR": VaR_insample,
    "OOS_VaR": VaR_oos,
    "InSample_Underperf": under_insample,
    "OOS_Underperf": under_oos,
    "InSample_NegPrem": neg_insample,
    "OOS_NegPrem": neg_oos
})

print("\n===== 3.3 COMPARISON TABLE (In-Sample vs OOS) =====")
display(table_33)


===== 3.3 COMPARISON TABLE (In-Sample vs OOS) =====


InSample_VaR   OOS_VaR  InSample_Underperf  OOS_Underperf  \
dp        -0.061545 -0.063320           -0.018652      -0.018093   
ep        -0.064144 -0.047868           -0.019573      -0.015519   
three     -0.065216 -0.074456           -0.018332      -0.022612   

       InSample_NegPrem  OOS_NegPrem  
dp                    0            5  
ep                    0            0  
three                 1            8

<span style="color: blue;">

ANS:

The predictors show mild in-sample optimism, and their performance deteriorates slightly out of sample.
Among them:

ep is the most stable and least prone to generating negative premiums.

dp and especially the three-variable model suffer from noticeable OOS degradation, particularly in terms of the frequency of negative risk-premium predictions.